fix(core): add warning when using zoneless but zone.js is still loaded

crisbeto opened the pull request:
Users may be using zoneless, but are still loading Zone.js in which case they won't get the full benefits like reduced bundle size. These changes detect such a case and log a warning.

pkozlowski-opensource reviewed (changes requested):
LGTM (with one nit comment) but as @JeanMeche noted the negative error code suggest that there are docs for this error, while docs are not part of this PR. Either add docs or flip the negative sign on the error code.

crisbeto commented:
I've switched it to a positive error code.

dylhunn reviewed (approved):
reviewed-for: fw-core

AndrewKushnir commented on packages/core/src/change_detection/scheduling/zoneless_scheduling_impl.ts:
```diff
@@ -293,6 +293,16 @@ export class ChangeDetectionSchedulerImpl implements ChangeDetectionScheduler {
+
+  if ((typeof ngDevMode === 'undefined' || ngDevMode) && typeof Zone !== 'undefined' && Zone) {
+    const message = formatRuntimeError(
+      RuntimeErrorCode.ZONELESS_REMOVE_ZONE,
```
nit:
```suggestion
      RuntimeErrorCode.UNEXPECTED_ZONEJS_PRESENT_IN_ZONELESS_MODE,
```

AndrewKushnir commented on packages/core/src/change_detection/scheduling/zoneless_scheduling_impl.ts:
```diff
@@ -293,6 +293,16 @@ export class ChangeDetectionSchedulerImpl implements ChangeDetectionScheduler {
+    const message = formatRuntimeError(
+      RuntimeErrorCode.ZONELESS_REMOVE_ZONE,
+      `The application is using zoneless change detection, but is still loading Zone.js.` +
+        `Consider removing Zone.js to get the full benefits of zoneless.`,
```
```suggestion
        `Consider removing Zone.js import (typically located in the 'polyfills.ts' file) to get the full benefits of zoneless.`,
```

JeanMeche replied:
`polyfills.ts` isn't generated by the CLI anymore since a couple of versions. We could mention the `polyfills` entry in `angular.json` though.

AndrewKushnir replied:
> polyfills.ts isn't generated by the CLI anymore since a couple of versions

Where do we add the zone.js import? We can update the error message to use a different location.

JeanMeche replied:
In the `angular.json`  : `"polyfills: ['zone.js']`

crisbeto replied:
I intentionally kept this a bit vague, because there are many ways to load Zone.js. E.g. it could also be a random `script` tag on the page.

AndrewKushnir replied:
@crisbeto thanks for additional info 👍 Do we expect to have zone.js at the same location for the majority of CLI apps? If that's the case, here is an alternative proposal for the error message, feel free to use and update it as you see fit. 
```suggestion
        `Consider removing Zone.js to get full benefits of zoneless. In CLI-based applications, ` + 
        `zone.js reference is typically located in the "polyfills" section of the \`angular.json\` file.`,
```

AndrewKushnir commented on packages/core/src/change_detection/scheduling/zoneless_scheduling_impl.ts:
```diff
@@ -293,6 +293,16 @@ export class ChangeDetectionSchedulerImpl implements ChangeDetectionScheduler {
 export function provideExperimentalZonelessChangeDetection(): EnvironmentProviders {
   performanceMarkFeature('NgZoneless');
+
+  if ((typeof ngDevMode === 'undefined' || ngDevMode) && typeof Zone !== 'undefined' && Zone) {
```
Quick question: why do we also do `&& Zone` after checking a `typeof`?

crisbeto replied:
It's just in case `Zone` is defined but null. It's a bit of an edge case, but it doesn't hurt to have it since it'll be removed in production mode.

AndrewKushnir reviewed (approved):
LGTM (just 1 **non-blocking** comment) 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

crisbeto committed:
fix(core): add warning when using zoneless but zone.js is still loaded

Users may be using zoneless, but are still loading Zone.js in which case they won't get the full benefits like reduced bundle size. These changes detect such a case and log a warning.

thePunderWoman reviewed (approved):
reviewed-for: fw-core, public-api

dylhunn commented:
This PR was merged into the repository by commit ae0baa25228d679c7a4df6950676b39417ad2f1a.

spock123 commented:
Just FYI, @angular/fire does not work without Zones, so the only way to use that with zoneless, is to still load zoneJS in polyfills, but use the zoneless change detection. Removal of zoneJS from polyfills will lead to @angular/fire crashing out badly.

So for a while , until they get that fixed (if they do), this will give a warning for a lot of people trying out zoneless with @angular/fire.
